### PR TITLE
fix: exit non-zero when a fatal error shuts the server down

### LINKED ISSUE #1092: fix(test): Bun proof scripts exit 0 on a failed assertion
## Problem

Every `test/bun/*.mjs` proof script that boots a server through `startServer` exits **0 when its assertions fail** under Bun, so CI reports green on a real regression.

Verified with a minimal repro (boot `startServer`, then `assert.equal(1, 2)`):

```
node t5.mjs  -> exit=1   (correct)
bun  t5.mjs  -> exit=0   (swallowed)
```

Cause: `startServer` calls `installProcessHandlers` (`packages/server/src/listener-core.js` L388), which registers an `uncaughtException` handler that runs the graceful shutdown, and `makeShutdown` ends in `process.exit(0)` on a clean close (L426). On Bun a top-level assertion failure in the module body routes through that handler, so the shutdown's exit code wins. Node reports the module-evaluation rejection before the handler can exit, which is why only Bun is affected in practice. The scripts pass a `quiet` logger, so the logged error is swallowed too and the run is completely silent.

Impact: `.github/workflows/ci.yml` runs roughly fifteen of these directly (`run: bun test/bun/smoke.mjs`, `listener.mjs`, `compression.mjs`, `file-storage.mjs`, `app-boot.mjs`, `seed.mjs`, `blog-db.mjs`, `path-alias.mjs`, `timeouts.mjs`, `binding-prefixes.mjs`, `slot-ssr-parity.mjs`, and others) and trusts the step's exit code. Those steps currently cannot fail.

Found while adding `test/bun/forwarded-proto.mjs` in #1091: its counterfactual (revert the fix, expect red) came back exit 0.

Not affected: `node scripts/run-bun-tests.js`, which spawns `bun test <file>` (L119) and classifies on both status and the reported fail count, so the `*.test.mjs` wrappers DO fail correctly there. The hole is specific to the direct plain-script CI steps.

## Design / approach

Two candidate fixes, not mutually exclusive:

1. **Per-script (the shape #1091 already uses):** capture the failure, run cleanup, then `process.exit(1)` explicitly. Self-contained and obviously correct, but it must be repeated in every script and a new script can silently forget it.
2. **Systemic:** have the proof scripts opt out of the framework's process handlers, or give `startServer` a way not to install them (a test-only flag), or set `process.exitCode = 1` before the shutdown handler runs. Preferred, since it fixes every existing script at once and cannot be forgotten by the next one.

Whichever is chosen, add a guard so this cannot silently regress: a meta-test that runs a deliberately-failing proof script under Bun and asserts a non-zero exit.

## Implementation notes (for the implementing agent)

- **Where to look:** `packages/server/src/listener-core.js` `installProcessHandlers()` (L388) and `makeShutdown()` (L410 onward, the `process.exit(0)` at L426). The scripts themselves are `test/bun/*.mjs`; the CI steps are in `.github/workflows/ci.yml` around L104 to L204. The working counterexample to copy is `test/bun/forwarded-proto.mjs` (from #1091).
- **Landmines:**
  - Do NOT simply stop installing the handlers in production. They are the graceful-shutdown path (SIGINT/SIGTERM drain, the 10s force-exit) and are load-bearing for a real deploy.
  - `installProcessHandlers` is guarded by a `globalThis.__webjsProcHandlers` latch, so a second `startServer` in one process is a no-op. A per-test opt-out must account for that.
  - The scripts pass a `quiet` logger, so any fix that only logs remains invisible. The exit code is the contract.
  - Verify the fix on BOTH runtimes: node currently exits 1 and must keep doing so.
- **Invariants:** `packages/` stays plain `.js` + JSDoc. Node/Bun parity (#508) is the whole point of this test layer.
- **Tests:** a meta-test asserting a failing proof script exits non-zero under Bun. A counterfactual is inherent here (the test IS the counterfactual).

## Acceptance criteria

- [ ] A `test/bun/*.mjs` proof whose assertion fails exits non-zero under Bun
- [ ] Node keeps exiting non-zero for the same failure
- [ ] Graceful shutdown on SIGINT/SIGTERM is unchanged for a real server
- [ ] A guard test prevents this from silently regressing
- [ ] The existing proof scripts are covered (systemic fix) or individually updated

### PR DESCRIPTION
**Pull request by vivek7405**
Closes #1092

A crashed server used to tell its supervisor it had stopped cleanly. `makeShutdown` ended every clean drain with `process.exit(0)` no matter what started it, so the exit code carried no information about whether the process was dying from an error.

The process now exits `0` only when an operator asked for the stop, it drained cleanly, and nothing crashed on the way out. Anything else is a `1`.

## What was actually broken

Two things, one cause.

**Production.** systemd `Restart=on-failure`, a Docker restart policy, and a Railway or Kubernetes crash-loop detector all read only the exit code. A crashed instance reported success and stayed down.

**The whole `test/bun/` proof layer could not fail.** Each script boots a real app through `startServer`, which installs the `uncaughtException` handler, so a failed top-level assertion reached that handler and the shutdown's `exit(0)` won. The scripts pass a `quiet` logger, so the run was silent as well as green. CI runs roughly twenty of those scripts directly and trusts the step's exit code.

**Correction to the issue:** it reports Bun only and says node exits 1. Node exits 0 too. Verified before the fix on both runtimes with a script that boots `startServer` then asserts `1 === 2`:

```
node repro.mjs  -> exit=0
bun  repro.mjs  -> exit=0
```

Node surfaces an ESM module-body throw as `uncaughtException`, not as the module-evaluation rejection the issue assumed, so it lands on the same handler. That widens the blast radius: `node test/bun/app-boot.mjs` in the in-repo-app job was swallowing failures too.

## The bug that survived inside the first fix

Review caught that the first cut captured the exit code at entry while the re-entrancy guard drops every later call. So a SIGTERM that had begun draining pinned the code at `0`, and an uncaught exception arriving before the drain settled was discarded. The drain window is up to ten seconds wide with in-flight requests still running in it, so a request throwing mid-drain is an ordinary event, not an exotic race, and the process still reported a clean stop. That is the exact failure this issue is about, alive inside the fix for it. Reproduced (SIGTERM, then a fatal 20ms into a 150ms drain, exited `0`) and fixed: the code is read at exit time and upgrades only, so a fatal mid-drain wins and a later signal cannot launder a crash back into a success.

The same round caught that the unit harness stubs `process.exit` while `makeShutdown` armed a 10s hard-exit timer it never cleared, leaving every test with a live timer pointed at the real `process.exit(1)`. Harmless only because those tests run last. The timer is disarmed before exit now.

## Why the systemic fix over the per-script one

The issue offered both. The per-script shape (an explicit `process.exit(1)`, what #1091 used) is self-contained but must be repeated in every script and the next one can silently forget it, which is how this class of bug returns. Fixing `makeShutdown` fixes every existing script at once and cannot be forgotten. It is also correct production behaviour on its own merits, independent of the tests.

The signal handlers' one-argument contract is unchanged; only `installProcessHandlers`'s `onFatal` passes `{ fatal: true }`.

## Deliberately not changed

`unhandledRejection` still only logs. A rejection with no handler leaves the process alive rather than exiting 0, so it is a hang rather than a silent green, and making it set a failing exit code would mean a long-running server that survived one mishandled promise would report failure on a later clean SIGTERM. The existing comment in `dev.js` says this is deliberate, and this PR does not relitigate it.

## Test plan

- `test/bun/proof-exit-code.mjs`, a cross-runtime guard asserting three arms so it cannot pass by making everything fail: a failed assertion is non-zero, a passing script is 0, and SIGTERM against a live server is still 0. Children are spawned with `process.execPath`, so it holds under whichever runtime runs it. Runs FIRST in the Bun CI job, since it is what makes every step below it able to fail.
- `test/bun/proof-exit-code.test.mjs`, the `node --test` wrapper, so `npm test` proves the node path.
- Five unit tests in `packages/server/test/listener/listener-core.test.js` covering `makeShutdown` directly: the operator-signal zero, the fatal-with-a-clean-drain one, the fatal-mid-drain upgrade, the no-downgrade direction, and the failed drain.
- Counterfactual, proven at `9eb3b6ad`: reverting only the source reds the guard on BOTH runtimes and reds one unit test. It was measured at that commit and dates to it; the branch has since added the mid-drain fix, which carries its own tests.
- Local: node suite 3692 pass / 0 fail, Bun matrix 275 pass / 0 genuine fail, e2e 88 pass / 0 fail, browser 0 failed across Chromium / Firefox / WebKit, four-app dogfood boot check clean with no broken preloads.

## Docs

- `packages/server/src/listener-core.js`, the `makeShutdown` jsdoc.
- `packages/server/AGENTS.md`, the `listener-core.js` module-map row.
- `website/app/docs/deployment/page.ts`, the Graceful Shutdown section, which previously described the shutdown without saying what code it exits with.
- `.agents/skills/webjs/references/testing.md`, on writing a plain proof script that boots a server, since the exit code is what carries its verdict.
- `test/bun/forwarded-proto.mjs`, whose header described this bug as still live.

### CODE REVIEW

**Review by vivek7405 — COMMENTED** [review]
Went back over this one with fresh eyes and it needed a real pass, mostly on the prose rather than the mechanism. The three-line change in `makeShutdown` is the right shape: the drain logic is untouched, only the code it exits with moved, and routing it through an explicit `{ fatal: true }` from `onFatal` rather than sniffing the signal string keeps the signal handlers' one-argument contract intact.

What I got wrong was describing it. I wrote that the exit code reports why the process is going down "not whether the drain worked", and that is not what the code does: a rejected `closeServer()` and a drain still hanging at the 10s deadline both exit 1 no matter which signal started them. The unit test I added in this same PR asserts exactly that, so the docs and the tests disagreed with each other on the paragraph an operator reads before wiring `Restart=on-failure`. Restated in all three places to the only accurate version: 0 means an operator asked for the stop AND it drained cleanly.

The other real one is arm 2 of the new proof. Its comment claimed it would catch a fix that failed every script collaterally, which it could not do, because it calls the plain `close` thunk rather than `makeShutdown` and then pinned its own exit code with an explicit `process.exit(0)`. Dropped the pin and said plainly which arm covers what.

One I am not changing: the worry that the `bun test` wrapper will tip the 5s per-test timeout the way `compression.test.mjs` and `listener-overhead.test.mjs` do. Those two boot two servers in a single test with a deliberate 400ms stall; this one spawns three short-lived children and lands at 570ms across repeated runs, and the matrix job is already green on a real runner, where a timeout would have been reported as a genuine failure rather than passing quietly. Reasoning below on the thread.

File: website/app/docs/deployment/page.ts [hunk 1]
```diff
@@ -104,6 +104,7 @@ npm run start -- --port 8080</pre>
...
     </ol>
     <p>Unhandled promise rejections are logged but do not crash the process. Uncaught exceptions trigger an orderly shutdown (state may be corrupted, so continuing is unsafe).</p>
+    <p><strong>The process exits <code>0</code> in one case only: an operator asked for the stop AND it drained cleanly.</strong> Everything else is a <code>1</code>. A drain that fails is a failure on its own terms, so a server close that errors, or a drain still hanging when the 10-second deadline fires, exits <code>1</code> whichever signal started it. When the drain succeeds, the code instead reports why the process is going down: <code>SIGINT</code> or <code>SIGTERM</code> exits <code>0</code>, while a shutdown started by an uncaught exception exits <code>1</code> even though the drain was perfectly clean, because the process is dying from an error. Your supervisor reads only that code, so this is what makes <code>Restart=on-failure</code> (systemd), a Docker restart policy, and a Railway or Kubernetes crash-loop detector actually fire on a crashed server instead of recording a successful exit and leaving the instance down.</p>
```
Comment (vivek7405) [inline website/app/docs/deployment/page.ts]: This says a signal-driven stop exits 0 and that the code does not reflect the drain, but a rejected `closeServer()` and the 10s hang both exit 1 regardless of the signal. The unit test added in this same PR asserts SIGTERM with a rejecting close exits 1, so this paragraph contradicts it, and it is the one an operator reads before wiring `Restart=on-failure`.

File: packages/server/src/listener-core.js [hunk 2]
```diff
@@ -405,14 +405,27 @@ export function installProcessHandlers(logger, onFatal) {
  * `closeServer()` thunk that resolves once the server has stopped accepting
  * connections (node `server.close`, Bun `server.stop(true)`). Closes the SSE hub,
  * then drains, then exits; hard-exits after 10s if the drain hangs.
+ *
+ * Exit code, in full. A drain that FAILS is always a failure, so a rejected
```
Comment (vivek7405) [inline packages/server/src/listener-core.js]: Same overclaim as the deployment docs: "not whether the drain itself worked" is not true of the failure paths just below it, where a hung or rejected drain exits 1 whatever started it.

File: test/bun/proof-exit-code.mjs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  // proof script does: close the server and let the runtime report the code.
+  // Note what this arm does and does NOT cover. `close` is the plain thunk
+  // `startServer` returns, NOT the `makeShutdown` handler, so this does not
+  // exercise the shutdown path at all; arm 3 is what covers `makeShutdown`'s
+  // zero path. What it catches is a fix to arm 1 that failed scripts
+  // COLLATERALLY (a stray non-zero `process.exitCode`, a close that starts
+  // reporting failure), which is the regression that would red all of CI while
+  // arm 1 stayed green. It is deliberately left without an explicit
+  // `process.exit(0)`, which would pin the code and make the arm vacuous.
+  const passing = w('passing.mjs', child(`assert.equal(1, 1);\nawait close();`));
```
Comment (vivek7405) [inline test/bun/proof-exit-code.mjs:134]: This arm cannot catch what the comment above it claims. `close` is the plain thunk `startServer` returns, not `makeShutdown`, and the explicit `process.exit(0)` pins the code before any shutdown path runs, so a regression making the shutdown always exit non-zero would still pass here. Arm 3 is the only one exercising `makeShutdown`'s zero path.

File: test/bun/forwarded-proto.mjs [hunk 4]
```diff
@@ -24,12 +24,15 @@
  * Run from the repo root so the bare `@webjsdev/server` specifier resolves.
  *
  * The failure is reported by an explicit `process.exit(1)` rather than by letting
- * the assertion propagate, because `startServer` installs an `uncaughtException`
- * handler that begins a graceful shutdown and exits 0. On Bun a top-level
- * assertion failure routes through that handler, so a broken proof would exit 0
- * and CI's `bun test/bun/<file>.mjs` step would go GREEN on a real regression
- * (verified: node exits 1, Bun exits 0). Filed separately as #1092 for the other
- * proof scripts, which all share this shape.
+ * the assertion propagate. That is now BELT AND BRACES rather than load-bearing:
```
Comment (vivek7405) [inline test/bun/forwarded-proto.mjs:27]: Still describes the swallowed exit code as current behaviour, and repeats the "node exits 1, Bun exits 0" claim that this PR disproves. This is the file cited as where the bug was found and the one a future agent copies the proof-script shape from.

File: test/bun/proof-exit-code.test.mjs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * plain assert script (`proof-exit-code.mjs`, not `*.test.mjs`) so the runner
+ * does not double-run it.
+ *
+ * This is the meta-test for the rest of `test/bun/`: it is what keeps every
+ * OTHER script in this directory able to fail.
+ */
+import { test } from 'node:test';
+
+test('a failing proof script exits non-zero on this runtime (#1092)', async () => {
+  await import('./proof-exit-code.mjs');
```
Comment (vivek7405) [inline test/bun/proof-exit-code.test.mjs:15]: This runs under the Bun matrix with `bun test`'s 5s default per-test timeout, and the single `test()` spawns three children that each boot a real app. `compression.test.mjs` and `listener-overhead.test.mjs` are both denylisted for that exact class. Does this need a DENYLIST entry too?

File: website/app/docs/deployment/page.ts (hunk 1 above)
Comment (vivek7405) [reply website/app/docs/deployment/page.ts]: Fixed in [`617843f6`](https://github.com/webjsdev/webjs/commit/617843f6). Now states the only accurate version: 0 means an operator asked for the stop AND it drained cleanly, everything else is 1.

File: packages/server/src/listener-core.js (hunk 2 above)
Comment (vivek7405) [reply packages/server/src/listener-core.js]: Fixed in [`617843f6`](https://github.com/webjsdev/webjs/commit/617843f6), same rewrite as the deployment docs so the JSDoc and the page cannot drift apart again.

File: test/bun/proof-exit-code.mjs (hunk 3 above)
Comment (vivek7405) [reply test/bun/proof-exit-code.mjs:134]: Fixed in [`617843f6`](https://github.com/webjsdev/webjs/commit/617843f6). Dropped the explicit `process.exit(0)` so a stray failing code surfaces, and rewrote the comment to say what the arm actually covers (collateral damage to a passing script) and that arm 3 is what covers `makeShutdown`'s zero path.

File: test/bun/forwarded-proto.mjs (hunk 4 above)
Comment (vivek7405) [reply test/bun/forwarded-proto.mjs:27]: Fixed in [`617843f6`](https://github.com/webjsdev/webjs/commit/617843f6). Rewrote it to say the explicit exit is belt and braces now rather than load-bearing, and corrected the node half of the claim.

File: test/bun/proof-exit-code.test.mjs (hunk 5 above)
Comment (vivek7405) [reply test/bun/proof-exit-code.test.mjs:15]: Not adding one. The two denylisted precedents boot two servers inside a single test with a deliberate 400ms stall; this spawns three short-lived children and measures 570ms, 570ms, 565ms across repeated runs against the 5s budget. The matrix already ran it green on a CI runner, and the runner classifies a timeout as a genuine failure rather than a pass, so a real overrun would have turned that job red instead of hiding. Denylisting it would also cost the node-path coverage that the wrapper exists to provide.

Comment (vivek7405) [reply test/bun/proof-exit-code.test.mjs:15]: Correcting one thing I said above: I lumped both precedents together as duration cases, and only one is. `listener-overhead.test.mjs` is denylisted for two server boots plus a deliberate 400ms stall in a single test, but `compression.test.mjs` is denylisted for error MIS-ATTRIBUTION (an intentional mid-stream ReadableStream error the bun runner pins on the wrong test), which is not about time at all. So the shape here matches neither.

The rejection itself holds, and on firmer ground than my first measurement. Seventeen runs under the matrix's exact `bun test <file>` invocation land at 350 to 578ms, and pinned to one core with `taskset -c 0` the worst run is 748ms against the 5000ms budget. The per-test timeout does apply (a deliberate 6.5s probe fails at 5000ms), but `classify()` in the runner fails on non-zero exit AND on a non-zero fail count, so an overrun is a hard red rather than a quiet pass. The direct CI step ran ~1.39s vs 0.537s locally, so that runner is at most ~2.6x slower, leaving roughly 3.5x of headroom.

File: packages/server/src/listener-core.js [hunk 6]
```diff
@@ -405,12 +405,34 @@ export function installProcessHandlers(logger, onFatal) {
...
+  // The drain window is up to 10s wide and in-flight requests are still running
+  // in it, so an uncaught exception landing there is an ordinary event, not an
+  // exotic race. Because the re-entrancy guard below drops that second call, a
+  // code captured at entry would stay 0 and the crashed process would report a
+  // successful stop, which is the exact failure this whole mechanism exists to
+  // prevent. Upgrade only (1 never falls back to 0), so a signal arriving during
+  // a fatal shutdown cannot launder the crash into a success either.
+  let code = 0;
+  return (signal, { fatal = false } = {}) => {
+    if (fatal) code = 1;
```
Comment (vivek7405) [inline packages/server/src/listener-core.js:438]: The exit code used to be captured here at entry, and the re-entrancy guard on the next line drops any later call, so a fatal landing inside an already-running SIGTERM drain was discarded and the process exited 0. A crash during a graceful shutdown reporting success is the exact failure this PR is about.

File: packages/server/src/listener-core.js [hunk 7]
```diff
@@ -420,11 +442,16 @@ export function makeShutdown({ closeServer, hub, logger }) {
       process.exit(1);
     }, 10_000);
     if (typeof hard.unref === 'function') hard.unref();
+    // Disarm before exiting. In production `process.exit` ends everything so this
+    // is redundant, but it keeps the deadline from outliving the decision it was
+    // guarding, which is what let a stubbed-exit test leave a live timer armed to
+    // kill an unrelated process 10s later.
+    const done = (c) => { clearTimeout(hard); process.exit(c); };
```
Comment (vivek7405) [inline packages/server/src/listener-core.js:452]: The 10s hard-exit timer was never cleared. Harmless in production where `process.exit` ends everything, but the unit tests stub `process.exit`, so every one of them left a live timer armed to call the real `process.exit(1)` ten seconds later, with only a quiet-logger warning as evidence.

**Review by vivek7405 — COMMENTED** [review]
Second pass over the whole thing, and it turned up a real one rather than more prose: the bug this PR is about was still alive inside the fix for it.

`makeShutdown` captured the exit code at entry, and the re-entrancy guard drops every later call, so a SIGTERM that had begun draining pinned the code at 0 and a fatal arriving afterwards was thrown away. The drain window is up to ten seconds wide and in-flight requests are still running in it, so a request throwing mid-drain is an ordinary event, not an exotic race, and the process told systemd or Railway it had stopped cleanly. I reproduced it before touching anything (a SIGTERM, then a fatal 20ms into a 150ms drain, exited 0) and after (exits 1). Worse, the prose I had just added asserted this could not happen, so the docs were actively wrong in the direction that matters. The code is read at exit time now and upgrades only, so a later signal cannot launder a crash back into a success either; both directions have a test.

The other one is smaller but the same family. The unit harness stubs `process.exit`, and `makeShutdown` armed a 10s hard-exit timer it never cleared, so each test left a live timer pointed at the REAL `process.exit(1)`. It only failed to bite because those tests happen to run last in the file. Anything appended after them would have inherited a spurious exit 1 whose only trace was a warning the quiet logger ate, which is precisely the silent-exit-code class this PR exists to remove. The timer is disarmed before exit now.

Also worth recording: the earlier claim that this needed four prose fixes was itself wrong, there were five surfaces, and the one I had missed was the skill reference an agent reads before writing a new proof script. All of them now say the same thing.

File: packages/server/src/listener-core.js (hunk 6 above)
Comment (vivek7405) [reply packages/server/src/listener-core.js:438]: Fixed in [`fef7e41b`](https://github.com/webjsdev/webjs/commit/fef7e41b).

File: packages/server/src/listener-core.js (hunk 7 above)
Comment (vivek7405) [reply packages/server/src/listener-core.js:452]: Fixed in [`fef7e41b`](https://github.com/webjsdev/webjs/commit/fef7e41b).

File: test/bun/proof-exit-code.mjs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  console.error(`FAIL proof-script exit codes on ${runtime}`);
+  console.error(failure instanceof Error ? failure.stack || failure.message : String(failure));
+  // Hard-exit ONLY as a standalone script (`node`/`bun test/bun/proof-exit-code.mjs`):
+  // Bun swallows a top-level-await rejection inside try/finally, so a plain throw
+  // would exit 0, which is the very failure this file exists to catch. When
+  // IMPORTED by the `.test.mjs` wrapper under `node --test`, a `process.exit(1)`
+  // would kill the whole single-process run and hide every other file's results,
+  // so throw instead and let the harness report one failed test. Same convention
+  // as `test/bun/listener-overhead.mjs`.
+  if (import.meta.main) process.exit(1);
```
Comment (vivek7405) [inline test/bun/proof-exit-code.mjs:169]: This was a bare `process.exit(1)`. The `.test.mjs` wrapper imports this file under `node --test`, where that kills the whole single-process run and hides every other file's results, so one failing proof would have erased the rest of the suite. `listener-overhead.mjs` documents the guarded form for exactly this case.

**Review by vivek7405 — COMMENTED** [review]
Last pass, and it found one that actually bites. The proof reported its own failure with a bare `process.exit(1)`. Standalone that is not just fine but necessary, because Bun swallows a top-level-await rejection inside try/finally and a plain throw would exit 0, which is the exact silent-green this whole PR is about. But the wrapper IMPORTS this file under `node --test`, and there an exit tears down the single-process run and takes every other file's results with it. So a single failing proof would have erased the rest of the suite's output, in a PR whose entire subject is test failures that fail to surface.

`listener-overhead.mjs` already documents the convention (`if (import.meta.main) process.exit(1); else throw failed`) and I simply had not followed it. Worse, the skill reference this PR adds was teaching the unguarded form to whoever writes the next proof script, so that is corrected too. Verified both directions: standalone with a broken assertion still exits 1 on node and on bun, and under `node --test` alongside another file it now reports 1 failed test while the other file still runs and passes.

The denylist question came up again and I am leaving it rejected. It was already put to a dedicated refuter that measured 17 runs at 350 to 578ms and a single-core worst case of 748ms against the 5s budget, and the Bun matrix has since gone green on every commit of this branch. The runner fails on a non-zero exit and on a non-zero fail count, so a timeout is a hard red rather than a quiet pass; if this file were tipping the budget the job would have said so by now.

Docs wording is settled across all five surfaces and I am not spending further rounds on prose.

File: test/bun/proof-exit-code.mjs (hunk 8 above)
Comment (vivek7405) [reply test/bun/proof-exit-code.mjs:169]: Fixed in [`9987682a`](https://github.com/webjsdev/webjs/commit/9987682a). Counterfactual both ways: standalone with a broken assertion still exits 1 on node and bun, and under `node --test` beside another file it reports 1 failed test while the other file still runs and passes (tests 2, pass 1, fail 1). Before the guard that run would have been killed outright.

**Comment by vivek7405** [comment]
**Design rationale: why the exit code upgrades instead of taking the last writer**

Two decisions here that the diff shows but does not explain, both worth having written down before someone "simplifies" them later.

**Why `code` upgrades rather than being overwritten.** The natural-looking version is to let each call set the code, so the most recent caller wins. That is wrong in one direction: a fatal shutdown is already in flight, the operator's rolling deploy then sends SIGTERM into it, and the crash gets laundered into a clean stop right before the process exits. Upgrade-only makes the failure sticky, which is the safe asymmetry, because the cost of a false 1 is a supervisor restarting a process that was going down anyway, while the cost of a false 0 is a crashed instance staying dead. There is deliberately no path back to 0 once anything has set 1, including the drain-failure paths.

**Why arm 2 of the proof exists even though it never touches `makeShutdown`.** It calls the plain `close` thunk, so it proves nothing about the shutdown path, and that is the point: it is the collateral-damage arm. The failure mode it guards is a future "fix" to arm 1 that makes scripts exit non-zero too eagerly, which would red every one of the ~20 direct CI steps at once while arm 1 stayed green and looked correct. Arm 3 is the arm that actually exercises `makeShutdown`'s zero path. Keeping all three is what makes the guard prove the mechanism discriminates rather than just fails loudly. Arm 2 is also deliberately left without an explicit `process.exit(0)`, which would pin its code and make it vacuous.

**One boundary left alone on purpose.** `unhandledRejection` still only logs. Making it set a failing exit code would mean a long-running server that survived a single mishandled promise would then report failure on an ordinary SIGTERM months later. A rejection with no handler leaves the process alive rather than exiting 0, so it surfaces as a hang, which is bad but not the silent-green class this change is about.

**The counterfactual is dated to its commit.** "Reverting the source reds the guard on both runtimes" was measured at the first commit of this branch. The mid-drain fix landed later and carries its own tests (the upgrade and the no-downgrade direction), so re-read that claim as being about the original three-line change, not about the branch head.
